lib/SILGen/SILGenPoly.cpp

```diff
@@ -3397,112 +3399,274 @@ SILGenFunction::getOrCreateAutoDiffLinearMapReorderingThunk(
+      // If result types are direct, add call result as direct thunk result.
+      if (toRes.isFormalDirect())
+        results.push_back(*fromDirResultsIter++);
+      // If result types are indirect, increment indirect result iterators.
+      else {
+        ++fromIndResultsIter;
+        ++toIndResultsIter;
+      }
+      continue;
     }
-    SmallVector<SILValue, 8> directResults;
-    auto tupleType = apply->getType().getAs<TupleType>();
-    if (tupleType && tupleType->getNumElements() != 0) {
-      for (auto i : range(1, tupleType->getNumElements()))
-        directResults.push_back(
-            thunkSGF.B.createTupleExtract(apply.getLoc(), apply, i));
-      directResults.push_back(
-          thunkSGF.B.createTupleExtract(apply.getLoc(), apply, 0));
-    } else {
-      directResults.push_back(apply);
+    // Load direct results from indirect results.
+    if (fromRes.isFormalIndirect()) {
+      auto indRes = *fromIndResultsIter++;
+      auto *load = thunkSGF.B.createLoad(
+          loc, indRes, LoadOwnershipQualifier::Unqualified);
+      results.push_back(load);
+      continue;
     }
-    auto reorderedResults =
-        joinElements(directResults, thunkSGF.B, apply.getLoc());
-    thunkSGF.B.createReturn(loc, reorderedResults);
-    break;
-  }
-  }
+    // Store direct results to indirect results.
+    assert(toRes.isFormalIndirect());
+    SILType resultTy = toConv.getSILType(toRes);
+    assert(resultTy.isAddress());
+    auto indRes = *toIndResultsIter++;
+    thunkSGF.B.createStore(loc, *fromDirResultsIter++, indRes,
+                           StoreOwnershipQualifier::Unqualified);
+  }
+  auto retVal = joinElements(results, thunkSGF.B, loc);
+
+  // Deallocate local allocations.
+  for (auto *alloc : reversed(localAllocations))
+    thunkSGF.B.createDeallocStack(loc, alloc);
+
+  // Create return.
+  thunkSGF.B.createReturn(loc, retVal);
   return thunk;
 }
 
+/// Forward function arguments, converting ownership.
+/// Adapted from `forwardFunctionArguments` in SILGenPoly.cpp.
+static void forwardFunctionArgumentsConvertingOwnership(
+    SILGenFunction &SGF, SILLocation loc, CanSILFunctionType fromTy,
+    CanSILFunctionType toTy, ArrayRef<ManagedValue> managedArgs,
+    SmallVectorImpl<SILValue> &forwardedArgs,
+    SmallVectorImpl<SILValue> &argumentsToFree) {
+  auto fromParameters = fromTy->getParameters();
+  auto toParameters = toTy->getParameters();
+  for (auto index : indices(managedArgs)) {
+    auto &arg = managedArgs[index];
+    auto fromParam = fromParameters[index];
+    auto toParam = toParameters[index];
+    // To convert owned argument to be guaranteed, retain the argument.
+    if (fromParam.isConsumed() && !toParam.isConsumed()) {
+      if (arg.getType().isObject())
+        SGF.B.createRetainValue(loc, arg.getValue(),
+                                SGF.B.getDefaultAtomicity());
+      else
+        SGF.B.createRetainValueAddr(loc, arg.getValue(),
```

Comment:
`retain_value_addr` can only be used on buffers of loadable types. I can't find any tests in this PR that involves abstraction differences in address-only parameters. If you write such an example I'm sure it will trigger an assertion.

When you convert an indirect address-only owned parameter (`@in`) to `@in_guaranteed`, you should allocate a new local buffer and perform `copy_addr` to create a copy of the underlying value. You can follow how this is done in the differentiation transform:
https://github.com/apple/swift/blob/b2720a3460c6eec5351e40f5fa2ddb95dd2df9a7/lib/SILOptimizer/Mandatory/Differentiation.cpp#L1956

Reply:
Parameter convention conversions for loadable-address and address-only types done in https://github.com/apple/swift/pull/26380/commits/6cbc71765c33fd4fde0d2b0c0235be5496c129e9, thanks!

I [added tests](https://github.com/apple/swift/pull/26380/commits/6cbc71765c33fd4fde0d2b0c0235be5496c129e9#diff-5240fec94d5989e4aacc603204b9ac0c) involving `__shared`/`__owned`/`__consuming` mismatches for struct methods/initializers. Is this sufficient coverage, or is more testing necessary (protocol methods with/without `__consuming`)?